### PR TITLE
ACS Adjustment Factor

### PR DESCRIPTION
**Pull request by nmrodelo**
Adding script to merge NHGIS ACS population by block group data to block level 2010 data aggregated to the block group level. 

Used to produce an adjustment factor to apply equally across block groups to replicate 2019 data.

### CODE REVIEW

File: data/merge_acs_data.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+# DATASETS : 	2015_2019_ACS5a
+# TOPICS : Total Population
+# The data should be downloaded at the block group level, the highest resolution geo data available from the ACS
+
+# The second file is our master block data set, which is produced elsewhere in this repository.
+# See 'create_master_block_dataset.py'
+
+
+acs = pd.read_csv(
+    r"C:\UBICenter\local-child-allowance\data\original\nhgis\nhgis0002_ds244_20195_2019_blck_grp.csv",
```
Comment (MaxGhenis) [inline data/merge_acs_data.py:15]: Read from a relative path, i.e. starting with `"data/`, and use forward slashes.

Comment (MaxGhenis) [inline data/merge_acs_data.py:15]: Is the data too big to be committed? If so, what about after gzipping?

File: data/merge_acs_data.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        "NAME_E",
+        "ALUBE001",
+        "ALUBM001",
+    ],
+)
+
+# low_memory = False is necessary to avoid an error message when loading in the huge block-level data
+
+blocks = pd.read_csv(
+    r"C:\UBICenter\local-child-allowance\data\master_block.csv.gz",
```
Comment (MaxGhenis) [inline data/merge_acs_data.py:31]: same thing, relative path

File: data/merge_acs_data.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+# The data can be found with these filters:
+# DATASETS : 	2015_2019_ACS5a
+# TOPICS : Total Population
+# The data should be downloaded at the block group level, the highest resolution geo data available from the ACS
+
+# The second file is our master block data set, which is produced elsewhere in this repository.
+# See 'create_master_block_dataset.py'
+
+
+acs = pd.read_csv(
```
Comment (MaxGhenis) [inline data/merge_acs_data.py:14]: name this `block_group`, or `acs_block_group` to distinguish from `census_block_group` later (clearer to specify the unit of the data in the dataset name)

File: data/merge_acs_data.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    low_memory=False,
+)
+
+
+# Fill missing values that interfere with merge with 0
+
+blocks.fillna(0)
+
+
+# The blockgroup is the same as the first digit of the four digit block number. Therefore, we can obtain it by integer dividing
```
Comment (MaxGhenis) [inline data/merge_acs_data.py:41]: check line length

File: data/merge_acs_data.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+# 1000
+
+blocks["block_group"] = (blocks["block"]) // 1000
+
+
+# Here we aggregate our block level data up to the block group level, so that we can then join it with the ACS data,
+# which is also at the block group level
+
+block_group = (
+    blocks.groupby(["state_fip", "county_fip", "census_tract", "block_group"])[
```
Comment (MaxGhenis) [inline data/merge_acs_data.py:51]: make the groupby variables a constant list so you can reuse it for the merge

File: data/merge_acs_data.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+# Fill missing values that interfere with merge with 0
+
+blocks.fillna(0)
+
+
+# The blockgroup is the same as the first digit of the four digit block number. Therefore, we can obtain it by integer dividing
+# 1000
+
+blocks["block_group"] = (blocks["block"]) // 1000
```
Comment (MaxGhenis) [inline data/merge_acs_data.py:44]:
```suggestion
blocks["block_group"] = blocks.block // 1000
```
also elsewhere, use dot notation when accessing a single column for consistency with the repo

File: data/merge_acs_data.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+# Finding the percentage change in population in each block group from 2010 to 2019 using ACS data
+
+acs_block_data["block_group_adjustment_factor"] = (
+    acs_block_data["acs_population"] / acs_block_data["population"]
+)
+
+
+# Using integer division once again to find the block group of a given block
+
+blocks["block_group"] = blocks["block"] // 1000
```
Comment (MaxGhenis) [inline data/merge_acs_data.py:94]: why do you have to do this again?

File: data/merge_acs_data.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+)
+
+
+# Finding block-level adjusted population using the adjustment factor from the ACS 2019 data
+
+adjusted_blocks["block_population_adj_2019"] = (
+    adjusted_blocks["block_population_2010"]
+    * adjusted_blocks["block_group_adjustment_factor"]
+)
+
```
Comment (MaxGhenis) [inline data/merge_acs_data.py:125]: Could you add checks here? e.g. 
```
assert adjusted_blocks.block_population_adj_2019.sum() > adjusted_blocks.block_population_2010.sum()
assert adjusted_blocks.block_population_adj_2019.sum().between(325e6, 330e6)
```

**Review by MaxGhenis — CHANGES_REQUESTED** [review]

File: data/merge_acs_data.py (hunk 1 above)
Comment (nmrodelo) [reply data/merge_acs_data.py:15]: yeah, too big to be committed, even after zipping. not sure why

File: data/merge_acs_data.py (hunk 7 above)
Comment (nmrodelo) [reply data/merge_acs_data.py:94]: to get the first digit of the 4 digit block id, which is also the block group number. this lets me do it without converting the data back and forth between string and int

File: data/merge_acs_data.py [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+    adjusted_blocks.block_pop_2010
+    * adjusted_blocks.block_group_adjustment_factor
+)
+
+assert (
+    adjusted_blocks.block_pop_adj_2019.sum()
+    > adjusted_blocks.block_pop_2010.sum()
+)
+
+assert (
```
Comment (MaxGhenis) [inline data/merge_acs_data.py:118]: duplicate

File: data/merge_acs_data.py [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+)
+
+# low_memory = False is necessary to avoid an error message when loading in the huge block-level data
+
+blocks = pd.read_csv("master_block.csv.gz", low_memory=False,)
+
+
+# Fill missing values that interfere with merge with 0
+
+blocks.fillna(0)
```
Comment (MaxGhenis) [inline data/merge_acs_data.py:36]: for this to do something it'll need `, inplace=True`

File: data/merge_acs_data.py [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+
+# Merging the acs block group level data back onto the more detailed block level data we started with.
+# This will allow us to apply the adjustment factor that we found above to all of the blocks within a block group
+
+adjusted_blocks = pd.merge(acs_block_data, blocks, on=merge_list, how="outer",)
+
+
+# Renaming for simplicity
+
+adjusted_blocks = adjusted_blocks.rename(
```
Comment (MaxGhenis) [inline data/merge_acs_data.py:97]: use `adjusted_blocks.rename(..., inplace=True)` instead

**Review by MaxGhenis — CHANGES_REQUESTED** [review]

File: data/merge_acs_data.py [hunk 12]
```diff
@@ -0,0 +1,130 @@
+import pandas as pd
+import numpy as np
+from pandas.core.reshape.merge import merge
```
Comment (MaxGhenis) [inline data/merge_acs_data.py:3]: remove this and numpy, as they're unused (fyi vscode lightens the font on unused imports)

File: data/merge_acs_data.py [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+
+# Using integer division once again to find the block group of a given block
+
+blocks["block_group"] = blocks.block // 1000
+
+
+# Merging the acs block group level data back onto the more detailed block level data we started with.
+# This will allow us to apply the adjustment factor that we found above to all of the blocks within a block group
+
+adjusted_blocks = pd.merge(acs_block_data, blocks, on=merge_list, how="outer",)
```
Comment (MaxGhenis) [inline data/merge_acs_data.py:92]: optionally, a bit more concise is:
```suggestion
adjusted_blocks = acs_block_data.merge(blocks, on=merge_list, how="outer")
```

File: data/merge_acs_data.py [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+# Therefore, we can obtain it by integer dividing by 1000
+
+blocks["block_group"] = blocks.block // 1000
+
+
+# Here we aggregate our block level data up to the block group level, so that we can then join it with the ACS data,
+# which is also at the block group level
+
+
+merge_list = ["state_fip", "county_fip", "census_tract", "block_group"]
```
Comment (MaxGhenis) [inline data/merge_acs_data.py:49]: constant are uppercase in python, "columns" is more precise
```suggestion
MERGE_COLS = ["state_fip", "county_fip", "census_tract", "block_group"]
```

**Review by MaxGhenis — CHANGES_REQUESTED** [review]

File: data/merge_acs_data.py (hunk 11 above)
Comment (nmrodelo) [reply data/merge_acs_data.py:97]: this broke my code for some reason... works fine without it. if I include this the next line afterwards throws an error about the variable not existing.

**Pull request merged**